description.md

```diff
@@ -0,0 +1,937 @@
+    - `DutchAuctionCalculator` - calculates amounts based on the time passed since the Dutch auction start. The rate is decreasing proportionally to the time passed.
+    - `RangeAmountCalculator` - calculates amounts based on the volumes filled and requested. The exchange rate is changing in the price range from minimum to maximum based on the volumes filled.
+
+**Order execution conditions**
+
+- `Predicate` - a predicate to validate the order against specified conditions at the moment of fill.
+
+**Permits and approvals**
+
+- `MakerPermit` - a maker’s permit to avoid getting approval for a token spend.
+
+**Interactions**
+
+- `PreInteractionData` and `PostInteractionData` - maker-defined interactions executed before and after assets transfer from maker to taker, and from taker to maker, respectively.
+
+**Custom data**
+
+- `CustomData` can be utilized by protocol extensions and is also passed to `takerInteraction` as part of extension calldata.
+
+### Extensions structure
+
+Extensions have a dynamic length and should be packed in the following structure:
+
+- The first 32 bytes of the calldata contain offsets for the extensions calldata. The offset for a specific parameter is coded as the offset of the end of the parameter's calldata. The offset of the start of the calldata is either the offset of the previous parameter, or zero for the first one.
+- Then, the extensions calldata follows correspondingly.
+
+Offsets contain a zero-based offset in calldata for each parameter and are packed as follows:
+
+| Parameter | Location, bytes |
+| --- | --- |
+| MakerAssetSuffix | [0..3] |
+| TakerAssetSuffix | [4..7] |
+| MakingAmountGetter | [8..11] |
+| TakingAmountGetter | [12..15] |
+| Predicate | [16..19] |
+| MakerPermit | [20..23] |
+| PreInteractionData | [24..27] |
+| PostInteractionData | [28..31] |
+
+The `CustomData` calldata is located after all extensions. Its start is defined as the end offset for `PostInteractionData`, and its end is the offset of the end of the calldata.
+
+**Example**
+
+For orders that have a predicate with a length of 120 bytes, a permit of 32 bytes included, and 32 bytes of custom data, the extension calldata should be packed as follows:
+
+| Offsets |  | MakerPermit offset | Predicate offset |  |  |  |  | Predicate calldata | Permit calldata | Custom data calldata |
+| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
+| [28..31] | [24..27] | [20..23] | [16..19] | [12..15] | [8..11] | [4..7] | [0..3] | [32..151] | [152..183] | [184..216] |
+| 152 | 152 | 152 | 120 | 0 | 0 | 0 | 0 | calldata (120 bytes) | calldata (32 bytes) | calldata (32 bytes) |
+
+The following is the final calldata which includes the offsets.
+
+```bash
+#  28-31    24-27    20-23    16-19    12-15     8-11      4-7      0-3
+00000098 00000098 00000098 00000078 00000000 00000000 00000000 00000000
+#         predicate end => ^^^^^^^^ ^^^^^^^^ <= predicate start
```

Comment:
Above in line 191 it is said that the offset of predicate is in 16..19 bits

Reply:
fixed